core/src/main/scala/filodb.core/memstore/PartKeyLuceneIndex.scala

```diff
@@ -272,6 +289,38 @@ class PartKeyLuceneIndex(ref: DatasetRef,
     indexWriter.close()
   }
 
+  // TODO - most frequently used eviction needed
+  private val readerStateCache = mutable.WeakHashMap[(IndexReader, String), DefaultSortedSetDocValuesReaderState]()
+  def labelValuesEfficient(colFilters: Seq[ColumnFilter], startTime: Long, endTime: Long,
+                           colName: String, limit: Int = 100): Seq[String] = {
+    require(facetEnabled, "Faceting not enabled on this index; labelValuesEfficient cannot be called")
+    val labelValues = mutable.ArrayBuffer[String]()
+    withNewSearcher { searcher =>
+      val reader = searcher.getIndexReader
+      // the state object is expensive to create, so reuse across queries. It is associated with reader
+      try {
+        val state = readerStateCache.getOrElseUpdate((reader, colName),
+            new DefaultSortedSetDocValuesReaderState(reader, FACET_FIELD_PREFIX + colName))
+        val fc = new FacetsCollector
+        val query = colFiltersToQuery(colFilters, startTime, endTime)
+        FacetsCollector.search(searcher, query, 10, fc)
+        val facets = new SortedSetDocValuesFacetCounts(state, fc)
+        val result = facets.getTopChildren(limit, colName)
+        if (result != null) {
+          for (i <- 0 until Math.min(result.childCount, limit)) {
```

Comment:
doesn't topk make sure result count has an upper bound of `limit`?

Reply:
I have changed this to `result.labelValues.length`